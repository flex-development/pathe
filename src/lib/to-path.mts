/**
 * @file toPath
 * @module pathe/lib/toPath
 */

import canParseURL from '#internal/can-parse-url'
import validateURLString from '#internal/validate-url-string'
import fileURLToPath from '#lib/file-url-to-path'
import type { ToPathOptions } from '@flex-development/pathe'

export default toPath

/**
 * Convert `input` to a path.
 *
 * @see {@linkcode ToPathOptions}
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {URL | string} input
 *  The {@linkcode URL}, URL string, or path to convert
 * @param {ToPathOptions | null | undefined} [options]
 *  Conversion options
 * @return {string}
 *  `input` as path
 */
function toPath(
  this: void,
  input: URL | string,
  options?: ToPathOptions | null | undefined
): string

/**
 * Convert a list of inputs to paths.
 *
 * @see {@linkcode ToPathOptions}
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {ReadonlyArray<URL | string>} list
 *  The list of {@linkcode URL}s, URL strings, or paths to convert
 * @param {ToPathOptions | null | undefined} [options]
 *  Conversion options
 * @return {string[]}
 *  List of paths
 */
function toPath(
  this: void,
  list: readonly (URL | string)[],
  options?: ToPathOptions | null | undefined
): string[]

/**
 * Convert inputs to paths.
 *
 * @see {@linkcode ToPathOptions}
 *
 * @category
 *  utils
 *
 * @this {void}
 *
 * @param {ReadonlyArray<URL | string> | URL | string} value
 *  The {@linkcode URL}, URL string, or path to convert, or list of such values
 * @param {ToPathOptions | null | undefined} [options]
 *  Conversion options
 * @return {string[] | string}
 *  `value` as path or new list of paths
 */
function toPath(
  this: void,
  value: readonly (URL | string)[] | URL | string,
  options?: ToPathOptions | null | undefined
): string[] | string

/**
 * @this {void}
 *
 * @param {ReadonlyArray<URL | string> | URL | string} value
 *  The {@linkcode URL}, URL string, or path to convert, or list of such values
 * @param {ToPathOptions | null | undefined} [options]
 *  Conversion options
 * @return {string[] | string}
 *  `value` as path or new list of paths
 */
function toPath(
  this: void,
  value: readonly (URL | string)[] | URL | string,
  options?: ToPathOptions | null | undefined
): string[] | string {
  if (Array.isArray<URL | string>(value)) {
    return value.map((input, i) => {
      return validateURLString(input, `value[${i}]`), toPath(input)
    })
  }

  validateURLString(value, 'input')

  if (typeof value === 'string') {
    if (!canParseURL(value)) return value
    value = new URL(value)
  }

  if (value.protocol === 'file:') return fileURLToPath(value, options)

  return value.pathname
}
