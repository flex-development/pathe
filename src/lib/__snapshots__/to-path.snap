// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:lib/toPath > should return \`input\` as path (".") 1`] = `"."`;

exports[`unit:lib/toPath > should return \`input\` as path ("../to-path.mts") 1`] = `"../to-path.mts"`;

exports[`unit:lib/toPath > should return \`input\` as path ("/") 1`] = `"/"`;

exports[`unit:lib/toPath > should return \`input\` as path ("/package.json") 1`] = `"/package.json"`;

exports[`unit:lib/toPath > should return \`input\` as path ("\\\\") 1`] = `"\\"`;

exports[`unit:lib/toPath > should return \`input\` as path ("file:///") 1`] = `"/"`;

exports[`unit:lib/toPath > should return \`input\` as path ("file:///package.json") 1`] = `"/package.json"`;

exports[`unit:lib/toPath > should return \`input\` as path ("node:test") 1`] = `"test"`;

exports[`unit:lib/toPath > should return \`input\` as path ("t:\\\\package.json") 1`] = `"t:\\package.json"`;
